### PR TITLE
fix(tracing_utils): resolves RuntimeError dictionary changed size during iteration

### PR DESCRIPTION
**Pull request by ggiill**
We've been seeing this infrequently in a FastAPI app and decided to dig in.

Looks like `context._meta` is changing size during iteration.

Saw a similar issue [here](https://github.com/DataDog/dd-trace-py/pull/1745) so implementing a similar fix.

---

`1.x` PR: https://github.com/DataDog/dd-trace-py/pull/7439

## Checklist

- [x] Change(s) are motivated and described in the PR description.
- [x] Testing strategy is described if automated tests are not included in the PR.
- [x] Risk is outlined (performance impact, potential for breakage, maintainability, etc).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] [Library release note guidelines](https://ddtrace.readthedocs.io/en/stable/releasenotes.html) are followed. If no release note is required, add label `changelog/no-changelog`.
- [x] Documentation is included (in-code, generated user docs, [public corp docs](https://github.com/DataDog/documentation/)).
- [x] Backport labels are set (if [applicable](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting))

## Reviewer Checklist

- [x] Title is accurate.
- [x] No unnecessary changes are introduced.
- [x] Description motivates each change.
- [x] Avoids breaking [API](https://ddtrace.readthedocs.io/en/stable/versioning.html#interfaces) changes unless absolutely necessary.
- [x] Testing strategy adequately addresses listed risk(s).
- [x] Change is maintainable (easy to change, telemetry, documentation).
- [x] Release note makes sense to a user of the library.
- [x] Reviewer has explicitly acknowledged and discussed the performance implications of this PR as reported in the benchmarks PR comment.
- [x] Backport labels are set in a manner that is consistent with the [release branch maintenance policy](https://ddtrace.readthedocs.io/en/latest/contributing.html#backporting)
- [x] If this PR touches code that signs or publishes builds or packages, or handles credentials of any kind, I've requested a review from `@DataDog/security-design-and-guidance`.
- [x] This PR doesn't touch any of that.

### CODE REVIEW

**Comment by ggiill** [comment]
@ZStriker19 @mabdinur @majorgreys bump - thanks!

**Comment by brettlangdon** [comment]
@ggiill thanks! any chance you were able to repro? any ideas on how we can create a regression test for this?

Would you be able to add a release note for this? https://github.com/DataDog/dd-trace-py/blob/306a8908107c33cc15e955d8fe773646d9870df8/docs/releasenotes.rst

**Comment by ggiill** [comment]
> @ggiill thanks! any chance you were able to repro? any ideas on how we can create a regression test for this?
> 
> Would you be able to add a release note for this? https://github.com/DataDog/dd-trace-py/blob/306a8908107c33cc15e955d8fe773646d9870df8/docs/releasenotes.rst

Sure - will add release notes!

Not sure exactly how to reproduce, but this is in a FastAPI (`AnyIO`/`asyncio`) app, so I assume something like this is happening:

```python
import asyncio

context_meta = {"a": 1, "b": 2}

async def f():
    for k, v in context_meta.items():
        print(f"f() | k: {k} | v: {v}")
        await asyncio.sleep(0.1)

async def g():
    for k, v in context_meta.items():
        k_to_add, v_to_add = k + "+", v + 1
        print(f"g() | adding | k: {k_to_add} | v: {v_to_add}")
        context_meta[k_to_add] = v_to_add
        await asyncio.sleep(0.1)

async def main():
    await asyncio.gather(f(), g())

asyncio.run(main())
```

Results in:
```
f() | k: a | v: 1
g() | adding | k: a+ | v: 2
Traceback (most recent call last):
  File "/Users/gil/example.py", line 24, in <module>
    asyncio.run(main())
  File "/Users/gil/.pyenv/versions/3.10.11/lib/python3.10/asyncio/runners.py", line 44, in run
    return loop.run_until_complete(main)
  File "/Users/gil/.pyenv/versions/3.10.11/lib/python3.10/asyncio/base_events.py", line 649, in run_until_complete
    return future.result()
  File "/Users/gil/example.py", line 21, in main
    await asyncio.gather(f(), g())
  File "/Users/gil/example.py", line 7, in f
    for k, v in context_meta.items():
RuntimeError: dictionary changed size during iteration
```

File: releasenotes/notes/fix-runtimeerror-on-get-metas-to-propagate-25f9c44ac4137fb9.yaml [hunk 1]
```diff
@@ -0,0 +1,5 @@
+---
+fixes:
+  - |
+    ddtrace.internal.utils: This fix resolves an issue where concurrent mutations to the `context._meta` dict
```
Comment (mabdinur) [inline releasenotes/notes/fix-runtimeerror-on-get-metas-to-propagate-25f9c44ac4137fb9.yaml:4]:
```suggestion
    tracing: This fix resolves an issue where concurrent mutations to the `context._meta` dict
```

**Review by mabdinur — DISMISSED** [review]

**Comment by mabdinur** [comment]
A regression test would be nice but it isn't a blocker

File: releasenotes/notes/fix-runtimeerror-on-get-metas-to-propagate-25f9c44ac4137fb9.yaml [hunk 2]
```diff
@@ -0,0 +1,5 @@
+---
+fixes:
+  - |
+    tracing: This fix resolves an issue where concurrent mutations to the `context._meta` dict
```
Comment (ZStriker19) [inline releasenotes/notes/fix-runtimeerror-on-get-metas-to-propagate-25f9c44ac4137fb9.yaml:4]:
```suggestion
    tracing: This fix resolves an issue where concurrent mutations to the ``context._meta`` dict
```

**Review by ZStriker19 — APPROVED** [review]

**Comment by ggiill** [comment]
@ZStriker19 Sorry - tried to merge but I guess it won't run the workflows/checks since I'm not in your org.

**Comment by ggiill** [comment]
@ZStriker19 - are you able to merge in 2.x, run the checks, and get this merged? 🙏

**Pull request merged**